### PR TITLE
[VAULT-4034] Revert back to caching nil values

### PR DESCRIPTION
**Pull request by pmmukh**
This reverts commit 67e1ed06c7199856f8493f416df55f06603881c6.

### CODE REVIEW

**Comment by swayne275** [comment]
why did we need to revert?

File: sdk/physical/cache.go [hunk 1]
```diff
@@ -184,10 +184,8 @@ func (c *Cache) Get(ctx context.Context, key string) (*Entry, error) {
 		return nil, err
 	}
 
-	if ent != nil {
-		// Cache the result
-		c.lru.Add(key, ent)
-	}
+	// Cache the result
```
Comment (ncabatoff) [inline sdk/physical/cache.go:187]:
```suggestion
	// Cache the result, even if nil
```

**Review by ncabatoff — APPROVED** [review]
Looks good, added comment suggestion so reader will know this is intentional.

**Comment by pmmukh** [comment]
@swayne275 so removing the nil caching was fixing this issue, but nil caching also solves a problem that brian discovered after some followups with jeff / research. Essentially, the nil caching ( caching that a value is absent ) was used as a performance optimization, to reduce calls to storage for missing data. We discussed this on a zoom a few minutes ago, and the plan going forward is to keep the nil caching, and let this issue be solved with cache invalidation. That cache invalidation isn't happening atm is planned to be fixed with this jira https://hashicorp.atlassian.net/browse/VAULT-4161

**Comment by swayne275** [comment]
@pmmukh thanks! how might we document that this is intended behavior?

**Comment by pmmukh** [comment]
@swayne275 np! so I think the suggestion nick added, should clarify that the nil caching is intentional

**Review by briankassouf — APPROVED** [review]

**Review by swayne275 — APPROVED** [review]

**Pull request merged**